src/randomness/RandomnessSender.sol

```diff
@@ -0,0 +1,147 @@
+import {ISignatureSender} from "../interfaces/ISignatureSender.sol";
+
+import {SignatureReceiverBase} from "../signature-requests/SignatureReceiverBase.sol";
+
+contract RandomnessSender is
+    IRandomnessSender,
+    SignatureReceiverBase,
+    Initializable,
+    UUPSUpgradeable,
+    AccessControlEnumerableUpgradeable
+{
+    // the DST is used to separate randomness being used as signatures for other things
+    string public constant DST = "randomness:0.0.1:bn254";
+    string public constant SCHEME_ID = "BN254";
+    bytes32 public constant ADMIN_ROLE = keccak256("ADMIN_ROLE");
+
+    uint256 public nonce = 0;
+
+    // Mapping from randomness requestID to callbacks
+    mapping(uint256 => TypesLib.RandomnessRequest) private callbacks;
+    // array of all requests
+    TypesLib.RandomnessRequest[] private allRequests;
+
+    event RandomnessRequested(
+        uint256 indexed requestID, uint256 indexed nonce, address indexed requester, uint256 requestedAt
+    );
+    event RandomnessCallbackSuccess(uint256 indexed requestID, bytes32 randomness, bytes signature);
+    event SignatureSenderUpdated(address indexed signatureSender);
+
+    error RandomnessCallbackFailed(uint256 requestID);
+
+    modifier onlyOwner() {
+        _checkRole(ADMIN_ROLE);
+        _;
+    }
+
+    /// @custom:oz-upgrades-unsafe-allow constructor
+    constructor() {
+        _disableInitializers();
+    }
+
+    function initialize(address _signatureSender, address owner) public initializer {
+        __UUPSUpgradeable_init();
+        __AccessControlEnumerable_init();
+
+        require(_grantRole(ADMIN_ROLE, owner), "Grant role failed");
+        require(_grantRole(DEFAULT_ADMIN_ROLE, owner), "Grant role failed");
+
+        require(_signatureSender != address(0), "Cannot set zero address as signature sender");
+        signatureSender = ISignatureSender(_signatureSender);
+    }
+
+    // OVERRIDDEN UPGRADE FUNCTIONS
+    function _authorizeUpgrade(address newImplementation) internal override onlyOwner {}
+
+    /**
+     * @dev See {IRandomnessSender-requestRandomness}.
+     */
+    function requestRandomness() external returns (uint256 requestID) {
+        nonce += 1;
+
+        TypesLib.RandomnessRequest memory r = TypesLib.RandomnessRequest({nonce: nonce, callback: msg.sender});
+        bytes memory m = messageFrom(r);
+        bytes memory conditions = hex"";
+
+        requestID = signatureSender.requestSignature(SCHEME_ID, m, conditions);
+
+        callbacks[requestID] = r;
+        allRequests.push(r);
+
+        emit RandomnessRequested(requestID, nonce, msg.sender, block.timestamp);
+    }
+
+    /**
+     * @dev See {SignatureReceiverBase-onSignatureReceived}.
+     */
+    function onSignatureReceived(uint256 requestID, bytes calldata signature) internal override {
+        TypesLib.RandomnessRequest memory r = callbacks[requestID];
+        require(r.nonce > 0, "request with that requestID did not exist");
```

Comment:
was gonna suggest checking against the current nonce too, but maybe it's fine to waste gas for people who put in crazy values